src/librustc_mir/interpret/memory.rs

```diff
@@ -941,3 +923,56 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         }
     }
 }
+
+/// A depth-first search over the allocation graph.
+///
+/// This is based on the DFS in `rustc_data_structures`, which we cannot use directly because
+/// `AllocId` does not implement `Idx`.
+pub struct DepthFirstSearch<'mem, 'mir, 'tcx, M: Machine<'mir, 'tcx>> {
+    memory: &'mem Memory<'mir, 'tcx, M>,
+    visited: FxHashSet<AllocId>,
+    stack: Vec<AllocId>,
+}
+
+impl<M: Machine<'mir, 'tcx>> DepthFirstSearch<'mem, 'mir, 'tcx, M> {
+    /// Returns a new DFS iterator that will traverse all allocations reachable from the given
+    /// `AllocId`s.
+    ///
+    /// The first node in `roots` will be the first node visited by the DFS.
+    pub fn with_roots(
+        memory: &'mem Memory<'mir, 'tcx, M>,
+        roots: impl IntoIterator<Item = AllocId>,
+    ) -> Self {
+        let mut stack: Vec<_> = roots.into_iter().collect();
+        stack.reverse();
+
+        DepthFirstSearch {
+            memory,
+            visited: stack.iter().copied().collect(),
+            stack,
+        }
+    }
+}
+
+impl<M: Machine<'mir, 'tcx>> Iterator for DepthFirstSearch<'mem, 'mir, 'tcx, M> {
+    type Item = (AllocId, InterpResult<'tcx, &'mem Allocation<M::PointerTag, M::AllocExtra>>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let DepthFirstSearch { stack, visited, memory } = self;
+
+        let id = stack.pop()?;
+        let alloc = memory.get_raw(id);
+
+        if let Ok(alloc) = alloc {
```

Comment:
Is this where you are ignoring `InterpError`? This seems reasonable.

We should make sure that we don't catch errors requiring allocation though, similar to [this](https://github.com/rust-lang/rust/blob/f5c81e0a986e4285d3d0fd781a1bd475753eb12c/src/librustc_mir/transform/const_prop.rs#L391). Maybe create a new helper method at the error type for this?

Reply:
I'm not sure what you mean by "errors requiring allocation". Are there certain errors that we should `bug` on instead of ignoring? Is `Ub` one of these?

Reply:
Any error type that has a `String` or other heap types in their fields will allocate when being thrown, and the catching will then destroy the allocated values. If this allocate->throw->catch->destroy happens in a high frequency we're wasting a lot of cycles with all the allocating/deallocating.

So, the idea is to make sure that any error that is caught and acted upon (and not just rethrown) does not allocate

Reply:
Idea: Should we make the set of errors that `Memory` can emit a separate type (with appropriate `Into` impls) so we know the set of errors that `Memory` methods can emit and can make sure all of them don't allocate?

Reply:
That makes sense. I'll probably wait for someone else to do this? It feels like y'all have a decent idea of what you want here.